homeassistant/components/qnap_qsw/config_flow.py

```diff
@@ -0,0 +1,73 @@
+"""Config flow for QNAP QSW."""
+from __future__ import annotations
+
+from typing import Any
+
+from aioqsw.exceptions import LoginError, QswError
+from aioqsw.localapi import ConnectionOptions, QnapQswApi
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_PASSWORD, CONF_URL, CONF_USERNAME
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers import aiohttp_client
+
+from .const import DOMAIN
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle config flow for a QNAP QSW device."""
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        errors = {}
+
+        if user_input is not None:
+            url = user_input[CONF_URL]
+            username = user_input[CONF_USERNAME]
+            password = user_input[CONF_PASSWORD]
+
+            qsw = QnapQswApi(
+                aiohttp_client.async_get_clientsession(self.hass),
+                ConnectionOptions(url, username, password),
+            )
+
+            serial = None
+            try:
+                system_board = await qsw.validate()
+            except LoginError:
+                errors["base"] = "invalid_auth"
+            except QswError:
+                errors["base"] = "cannot_connect"
+            else:
+                serial = system_board.get_serial()
+
+            if serial:
+                await self.async_set_unique_id(serial)
+                self._abort_if_unique_id_configured()
+            else:
+                self._async_abort_entries_match(
+                    {
+                        CONF_URL: url,
+                        CONF_USERNAME: username,
+                        CONF_PASSWORD: password,
+                    }
+                )
```

Comment:
Can this happen?

The title will end up being
"QNAP QSW None"

if it gets through here?

Reply:
I don't think this could happen at all, but since I've only tested one switch I prefer to guard against the possibility that the serial number isn't provided or is provided differently.
I will change the title to use the model instead.

Reply:
200fdffc6ad2abd9609f3e65f8f3ea709d61ba31

Reply:
Since its not expected to happen, we should abort if we can't get the serial number

Reply:
> Since its not expected to happen, we should abort if we can't get the serial number

@bdraco implemented in f3975085a80611b099edd34c3d672c73b3c2fb6c
I also decided to replace the serial number with the MAC address for several reasons:
1. The serial number returned by the API contains some weird characters that need to be filtered in order to have a printable serial number (https://github.com/Noltari/aioqsw/blob/38e15883b2fc00dd7596a70b29cf48664e9fa772/aioqsw/device.py#L347)
2. The serial number isn't displayed on the Switch website as opposed to the MAC address, which is displayed on several places in the Switch website. This way a user with multiple switches can easily distinguish them.

Reply:
Is it possible there are devices with multiple mac addresses (check `synology_dsm`)

Reply:
> Is it possible there are devices with multiple mac addresses (check `synology_dsm`)

As far as I know the local API just returns one MAC address, which is the one displayed on the website:
![image](https://user-images.githubusercontent.com/264346/164067802-009feffd-14ae-47ce-9c74-31825388dcf0.png)
![image](https://user-images.githubusercontent.com/264346/164067925-ed6e3c80-7ccc-499b-8b5b-43300f8bea27.png)